test/extended/openstack/machinesets.go

```diff
@@ -0,0 +1,145 @@
+		g.BeforeEach(func() {
+			g.By("preparing openshift dynamic client")
+			cfg, err := e2e.LoadConfig()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			dc, err = dynamic.NewForConfig(cfg)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			clientSet, err = e2e.LoadClientset()
+			o.Expect(err).NotTo(o.HaveOccurred())
+		})
+
+		g.It("should not have leftovers ports", func() {
+			// Check the scenario at https://bugzilla.redhat.com/show_bug.cgi?id=2073398
+
+			skipUnlessMachineAPIOperator(dc, clientSet.CoreV1().Namespaces())
+			g.By("Fetching worker machineSets")
+			var networkClient *gophercloud.ServiceClient
+			var rawBytes []byte
+			var newProviderSpec machineproviderv1.OpenstackProviderSpec
+			var rclient runtimeclient.Client
+
+			networkClient, err := client(serviceNetwork)
+			o.Expect(err).NotTo(o.HaveOccurred(), "Error creating an openstack network client")
+			machineSets, err := listWorkerMachineSets(dc)
+			o.Expect(err).NotTo(o.HaveOccurred(), "Error getting the workers machinesets")
+
+			if len(machineSets) == 0 {
+				e2eskipper.Skipf("Expects at least one worker machineset. Found none.")
+			}
+
+			cfg, err := e2e.LoadConfig()
+			o.Expect(err).NotTo(o.HaveOccurred(), "Error getting cluster config")
+
+			rclient, err = runtimeclient.New(cfg, runtimeclient.Options{})
+			o.Expect(err).NotTo(o.HaveOccurred(), "Error creating a runtime client")
+
+			err = machinev1.AddToScheme(scheme.Scheme)
+			o.Expect(err).NotTo(o.HaveOccurred(), "Failed to add Machine to scheme")
+			err = configv1.AddToScheme(scheme.Scheme)
+			o.Expect(err).NotTo(o.HaveOccurred(), "Failed to add Config to scheme")
+
+			newMachinesetParams := framework.BuildMachineSetParams(rclient, 1)
+			rawBytes, err = json.Marshal(newMachinesetParams.ProviderSpec.Value)
+
+			err = json.Unmarshal(rawBytes, &newProviderSpec)
+			o.Expect(err).NotTo(o.HaveOccurred(), "Error unmarshaling new MachineSet Provder Spec")
+
+			newProviderSpec.ServerGroupName = ""
+			newProviderSpec.ServerGroupID = "boo"
+			newMachinesetParams.Name = "bogus1-worker"
+			newProviderSpecJson, err := json.Marshal(newProviderSpec)
+			o.Expect(err).NotTo(o.HaveOccurred(), "Failed to marshal new Machineset provider spec")
+			newMachinesetParams.ProviderSpec.Value.Raw = newProviderSpecJson
+
+			g.By("Create a new machineSet with bogus server group ID")
+			ms, err := framework.CreateMachineSet(rclient, newMachinesetParams)
+			o.Expect(err).NotTo(o.HaveOccurred(), "Failed to create a Machineset")
+			defer DeleteMachinesetsDefer(rclient, ms)
+
+			clusterID := ms.Spec.Template.ObjectMeta.Labels["machine.openshift.io/cluster-api-cluster"]
+
+			const maxRetries = 5
+			const delay = 10
+			retries := 1
+			for retries < maxRetries {
+				_, err = framework.GetMachineSet(rclient, ms.Name)
+				if err != nil {
+					retries += 1
+					time.Sleep(time.Second * delay)
+				} else {
+					break
+				}
+			}
+			o.Expect(err).NotTo(o.HaveOccurred(), "Failed to get the new Machineset")
```

Comment:
if an error happened once and eventually the machine was found, will we still have an err here and fail when it shouldn't?

Reply:
If at the last iteration of the loop there was an error then it raises it , even there wasn't then it breaks out of the loop and the expect should be fine